### ISSUE TITLE
face_encodings method can not run on asynchronous threads

### ISSUE THREAD

**Issue by weili905906480** [ageitgey/face_recognition#374]
* face_recognition version: 
* Python version: 3.4
* Operating System: windows 7

### Description

I want to use face_recognition.face_encodings method on asynchronous threads, Because it causes my webcam to be unsmooth, but when I use it on asynchronous threads, It has no effect at all

```
class CompareByThreadPool(object):
    executor = ThreadPoolExecutor(max_workers=1)
   
    def __init__(self):

        self.mutex = QMutex()
        self.obama_img = []
        self.obama_face_encoding = []

        self.face_locations = []
        self.face_encodings = []
        self.face_names = []
        self.process_this_frame = False
        self.can_begain_compare = False
        self.can_begain_encoding = False
        self.frame=numpy.arange(0)
        future = self.executor.submit(self.cricleRun,self.frame)

    def set_frame(self,frame):
        self.frame = frame

    def cricleRun(self,frame):
        while True:
            print('self.frame.size: ',self.frame.size)
            if self.frame.size > 0:
                self.face_rec(self.frame)
            time.sleep(0.4)
    
    def face_rec(self,frame):
        
        small_frame = cv2.resize(frame, (0, 0), fx=0.25, fy=0.25)
    
        # Convert the image from BGR color (which OpenCV uses) to RGB color (which face_recognition uses)
        rgb_small_frame = small_frame[:, :, ::-1]
        # rgb_small_frame = small_frame
      
        self.face_locations = face_recognition.face_locations(rgb_small_frame)

        rec_begain_time = time.time()
        print('begain time: ',rec_begain_time)
        if self.process_this_frame:
            self.face_encodings = face_recognition.face_encodings(rgb_small_frame, self.face_locations)
        self.process_this_frame = not self.process_this_frame
        print('face_rec compareFlag: ',self.can_begain_compare)
        if self.can_begain_compare:
            face_names = []
            for face_encoding in self.face_encodings:
                #match = face_recognition.compare_faces([obama_face_encoding], face_encoding,0.5)
                face_distances = face_recognition.face_distance([self.obama_face_encoding],face_encoding);
                print("face_distance: ",face_distances[0]);
                #if match[0]:
                if face_distances[0] < 0.6:
                    name = "Success"
                else:
                    name = "unknown"
                rec_end_time= time.time()
                face_names.append(name)
                print("name : {0} time: {1} ".format(name,rec_end_time-rec_begain_time))
```

### What I Did

```
Can it run on asynchronous threads ， Please tell me how can I make my webcam smooth
```

**Comment by ageitgey** [comment]
There's several problems with your code. 

First, don't try to use multiprocessing inside of a class like this. Having class-level flags like `self.can_begain_compare` just isn't going to work in a multithreaded program with multiple threads accessing it at the same time. Instead, have a single function outside of a class that gets run by each thread and make that function take in a fixed set of inputs and return a fixed result. The code that runs in the thread should not access any global/class variables.

Second, you are only calling `executor.submit` once on an empty array, not on an image from the webcam. You aren't ever grabbing new images from the webcam or passing them to the executor from what I can see. And you aren't ever getting the result of the future that you asked it to process.

Third, using a thread pool probably won't work well inside of python anyway because of how it has a global interpreter lock across all threads. You probably won't gain that much real-world performance increase. You'd probably get a better result using a ProcessPool instead. This is just a guess though - I haven't tested it with a thread pool.

But even if you get that working, you have a problem of synchronizing the order of frames that you get back from the threads (assuming that you have multiple threads doing the face recognition step in parallel). The threads are going to finish out of order because different frames will need different amounts of processing. If you show each frame of video as a thread finishes processing it, your final video will jump around and the frames will be out of order. You need some way of keeping the results in order even if they are processed out of order. And just putting all the frames in a queue will cause "frame pacing" freezes in your video where the video speeds up and slows down in a weird way as different frames take more or less time to process by different threads.

In other words, using threads to process video can involve writing some fairly complicated synchronization code or coming up with a more complex design where different threads do different steps of processing (like one thread to do face location, one to do face encoding, etc). It's beyond the scope of something I can answer in a github issue.

**Comment by weili905906480** [comment]
Thank you very much for your reply ! I'm not a professional python engineer，I change to use c++ to implement  it already，Thank you very much at last. You are a good man! @ageitgey

**Comment by Mars212** [comment]
the same problem ,why face_recognition.face_encodings(rgb_small_frame, face_locations) can not run on asynchronous threads,  how to solve it
